grakn-graql/src/main/java/ai/grakn/graql/internal/reasoner/query/ReasonerAtomicQueryIterator.java

```diff
@@ -0,0 +1,122 @@
+ * along with Grakn. If not, see <http://www.gnu.org/licenses/gpl.txt>.
+ */
+
+package ai.grakn.graql.internal.reasoner.query;
+
+import ai.grakn.graql.admin.Answer;
+import ai.grakn.graql.admin.Unifier;
+import ai.grakn.graql.internal.query.QueryAnswer;
+import ai.grakn.graql.internal.reasoner.cache.QueryCache;
+import ai.grakn.graql.internal.reasoner.explanation.LookupExplanation;
+import ai.grakn.graql.internal.reasoner.explanation.RuleExplanation;
+import ai.grakn.graql.internal.reasoner.iterator.ReasonerQueryIterator;
+import ai.grakn.graql.internal.reasoner.rule.InferenceRule;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.Set;
+import java.util.stream.Stream;
+import javafx.util.Pair;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ *
+ * <p>
+ * Tuple-at-a-time iterator for atomic queries.
+ * Resolves the atomic query by:
+ * 1) doing DB lookup
+ * 2) applying a rule
+ * 3) doing a lemma (previously derived answer) lookup from query cache
+ * </p>
+ *
+ * @author Kasper Piskorski
+ *
+ */
+class ReasonerAtomicQueryIterator extends ReasonerQueryIterator {
+
+    private final ReasonerAtomicQuery query;
+
+    private final QueryCache<ReasonerAtomicQuery> cache;
+    private final Set<ReasonerAtomicQuery> subGoals;
+    private final Iterator<InferenceRule> ruleIterator;
+    private Iterator<Answer> queryIterator = Collections.emptyIterator();
+    private Unifier cacheUnifier = new UnifierImpl();
+
+    private InferenceRule currentRule = null;
+
+    private static final Logger LOG = LoggerFactory.getLogger(ReasonerAtomicQuery.class);
+
+    ReasonerAtomicQueryIterator(ReasonerAtomicQuery q, Answer sub, Set<ReasonerAtomicQuery> subGoals, QueryCache<ReasonerAtomicQuery> qc){
+        this.subGoals = subGoals;
+        this.cache = qc;
+        this.query = new ReasonerAtomicQuery(q);
+
+        query.addSubstitution(sub);
+
+        LOG.trace("AQ: " + query);
+        LOG.trace("AQ delta: " + sub);
+
+        Pair<Stream<Answer>, Unifier> streamUnifierPair = query.lookupWithUnifier(cache);
+        this.queryIterator = streamUnifierPair.getKey().iterator();
+        this.cacheUnifier = streamUnifierPair.getValue().invert();
+
+        //if this already has full substitution and exists in the db then do not resolve further
+        //NB: the queryIterator check is purely because we may want to ask for an explanation
+        boolean hasFullSubstitution = query.hasFullSubstitution();
+        if(subGoals.contains(query)
+                || (hasFullSubstitution && queryIterator.hasNext() ) ){
+            this.ruleIterator = Collections.emptyIterator();
+        }
+        else {
+            this.ruleIterator = query.getRuleIterator();
+        }
+
+        //mark as visited and hence not admissible
+        if (ruleIterator.hasNext()) subGoals.add(query);
+    }
+
+    @Override
+    public boolean hasNext() {
+        if (queryIterator.hasNext()) return true;
+        else{
```

Comment:
Same here, no need for the else block. 

As a future nitpick if you end up having lots of implementations of `hasNext` and if all the implementations have a similar structure then it maybe possible to get rid of all the boiler plate `ifs` by adding a common method in the super class to handle that boiler plate code. Something like:

```
public boolean hasNextHelper(Iterator it,  Supplier<Boolean> checker, Supplier<Boolean> finalCheck){
    if(it.hasNext()) return true;
    return checker.get() && finalCheck.get();
}
```

But that's a nitpick only if we see more of this.